br/pkg/lightning/backend/local/duplicate.go

```diff
@@ -180,645 +234,570 @@ func physicalTableIDs(tableInfo *model.TableInfo) []int64 {
+) error {
+	remainKeyRanges := newPendingKeyRanges(task.KeyRange)
+	err := common.Retry("process remote dup task", logger, func() error {
+		var (
+			regions   []*restore.RegionInfo
+			keyRanges []tidbkv.KeyRange
+		)
+		for _, kr := range remainKeyRanges.list() {
+			subRegions, subKeyRanges, err := m.splitKeyRangeByRegions(ctx, kr)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			regions = append(regions, subRegions...)
+			keyRanges = append(keyRanges, subKeyRanges...)
+		}
+
+		var metErr common.OnceError
+		wg := &sync.WaitGroup{}
+		for i := 0; i < len(regions); i++ {
+			region := regions[i]
+			kr := keyRanges[i]
+
+			wg.Add(1)
+			regionPool.Apply(func() {
+				defer wg.Done()
+
+				logger := logger.With(
+					zap.Uint64("regionID", region.Region.Id),
+					logutil.Key("dupDetectStartKey", kr.StartKey),
+					logutil.Key("dupDetectEndKey", kr.EndKey),
+				)
+				err := common.Retry("collect remote duplicate by region", logger, func() error {
+					stream, err := NewRemoteDupKVStream(region, kr, importClientFactory)
+					if err != nil {
+						return errors.Annotatef(err, "failed to create remote duplicate kv stream")
+					}
+					if task.indexInfo == nil {
+						err = m.RecordDataConflictError(ctx, stream)
+					} else {
+						err = m.RecordIndexConflictError(ctx, stream, task.tableID, task.indexInfo)
+					}
+					if err != nil {
+						return errors.Annotatef(err, "failed to record conflict errors")
+					}
+					return nil
+				})
+				if err != nil {
+					logger.Warn("[detect-dupe] collect duplicate rows from region failed", log.ShortError(err))
+					metErr.Set(err)
+				} else {
+					logger.Debug("[detect-dupe] collect duplicate rows from region completed")
+					remainKeyRanges.finish(kr)
+				}
+			})
 		}
-		reqs = append(reqs, req)
+		wg.Wait()
+		return metErr.Get()
+	})
+	return errors.Trace(err)
+}
+
+func (m *DuplicateManager) CollectDuplicateRowsFromTiKV(ctx context.Context, importClientFactory ImportClientFactory) error {
+	tasks, err := m.buildDupTasks()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	logger := m.logger
+	logger.Info("[detect-dupe] collect duplicate rows from tikv", zap.Int("tasks", len(tasks)))
+
+	pool := utils.NewWorkerPool(uint(m.regionConcurrency), "collect duplicate rows from tikv")
+	regionPool := utils.NewWorkerPool(uint(m.regionConcurrency), "collect duplicate rows from tikv by region")
+	g, gCtx := errgroup.WithContext(ctx)
+	for _, task := range tasks {
+		task := task
+		pool.ApplyOnErrorGroup(g, func() error {
+			taskLogger := logger.With(
+				logutil.Key("startKey", task.StartKey),
+				logutil.Key("endKey", task.EndKey),
+				zap.Int64("tableID", task.tableID),
+			)
+			if task.indexInfo != nil {
+				taskLogger = taskLogger.With(
+					zap.String("indexName", task.indexInfo.Name.O),
+					zap.Int64("indexID", task.indexInfo.ID),
+				)
+			}
+			err := m.processRemoteDupTask(gCtx, task, taskLogger, importClientFactory, regionPool)
```

Comment:
Share the same workerPool between these two level function may lead to dead lock if there are many dup tasks. This is likely happened when the table contains a lot of partitions or indices. I suggest use a separate pool the control this concurrency.

Reply:
`regionPool` is already a separate pool. Do you mean merge the `pool` and `regionPool`?